File: companion-apps/ios/CompanionApp/Integration/Diagnostics/RoutingDiagnosticsStore.swift
import Foundation

@MainActor
final class RoutingDiagnosticsStore: ObservableObject {
    @Published var sessions: [RoutingDiagSession] = []
    @Published var currentSession: RoutingDiagSession? = nil

    private let persistence: CompanionPersistence
    private var eventCounter = 0

    var isRecording: Bool { currentSession != nil }

    init(persistence: CompanionPersistence) {
        self.persistence = persistence
        self.sessions = persistence.loadRoutingDiagnosticsSessions()
    }

    func startRecording() {
        guard currentSession == nil else { return }
        currentSession = RoutingDiagSession(
            id: newSessionId(),
            createdAtMs: nowMs(),
            updatedAtMs: nowMs(),
            events: []
        )
    }

    func stopRecording() {
        guard let session = currentSession else { return }
        var finalized = session
        finalized.updatedAtMs = nowMs()
        currentSession = nil
        persistence.saveRoutingDiagnosticsSession(finalized)
        sessions = persistence.loadRoutingDiagnosticsSessions()
    }

    func recordEvent(_ data: RoutingDiagEventData) {
        guard currentSession != nil else { return }
        let event = RoutingDiagEvent(
            id: newEventId(&eventCounter),
            timestampMs: nowMs(),
            data: data
        )
        currentSession?.events.append(event)
        currentSession?.updatedAtMs = nowMs()
    }

    func recordRouteGeometry(
        routeId: String,
        providerName: String,
        geometry: [CoordinatePoint]
    ) {
        guard let session = currentSession else { return }
        let existing = session.routeGeometries ?? []
        guard !existing.contains(where: { $0.routeId == routeId }) else { return }
        let entry = RouteGeometryEntry(
            routeId: routeId,
            providerName: providerName,
            geometry: geometry
        )
        currentSession?.routeGeometries = existing + [entry]
    }

    func deleteSession(id: String) {
        persistence.dismissRoutingDiagnosticsSession(id: id)
        sessions = persistence.loadRoutingDiagnosticsSessions()
    }

    func debugPackageText(for id: String) -> String? {
        sessions.first(where: { $0.id == id })?.debugPackageText
    }

    private func nowMs() -> UInt64 {
        UInt64(Date().timeIntervalSince1970 * 1000)
    }
}

File: companion-apps/android/app/src/main/java/app/navon/bike/integration/diagnostics/RoutingDiagnosticsStore.kt
package app.navon.bike.integration.diagnostics

import app.navon.bike.domain.*
import app.navon.bike.integration.persistence.CompanionPersistence
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

class RoutingDiagnosticsStore(private val persistence: CompanionPersistence) {

    private val _sessions = MutableStateFlow<List<RoutingDiagSession>>(emptyList())
    val sessions: StateFlow<List<RoutingDiagSession>> = _sessions.asStateFlow()

    private val _currentSession = MutableStateFlow<RoutingDiagSession?>(null)
    val currentSession: StateFlow<RoutingDiagSession?> = _currentSession.asStateFlow()

    val isRecording: Boolean get() = _currentSession.value != null

    init {
        _sessions.value = persistence.loadRoutingDiagnosticsSessions()
    }

    fun startRecording() {
        if (isRecording) return
        _currentSession.value = RoutingDiagSession(
            id = newSessionId(),
            createdAtMs = nowMs(),
        )
    }

    fun stopRecording() {
        val session = _currentSession.value ?: return
        val finalized = session.copy(
            updatedAtMs = nowMs(),
            events = session.events.toList(),
        )
        _currentSession.value = null
        persistence.saveRoutingDiagnosticsSession(finalized)
        _sessions.value = persistence.loadRoutingDiagnosticsSessions()
    }

    fun recordEvent(data: RoutingDiagEventData) {
        _currentSession.update { session ->
            if (session == null) return
            val event = RoutingDiagEvent(
                id = newEventId(),
                timestampMs = nowMs(),
                data = data,
            )
            session.copy(
                updatedAtMs = nowMs(),
                events = session.events + event,
            )
        }
    }

    fun recordRouteGeometry(routeId: String, providerName: String, geometry: List<CoordinatePoint>) {
        _currentSession.update { session ->
            if (session == null) return
            val existing = session.routeGeometries ?: emptyList()
            if (existing.any { it.routeId == routeId }) return
            val entry = RouteGeometryEntry(
                routeId = routeId,
                providerName = providerName,
                geometry = geometry,
            )
            session.copy(routeGeometries = existing + entry)
        }
    }

    fun deleteSession(id: String) {
        persistence.dismissRoutingDiagnosticsSession(id)
        _sessions.value = persistence.loadRoutingDiagnosticsSessions()
    }

    fun debugPackageText(id: String): String? {
        return _sessions.value.find { it.id == id }?.debugPackageText()
    }
}
